### PR TITLE
Introducing the maximum size preview

### PR DESCRIPTION
**Pull request by oparoz**
#### Fixes
- Always store the preview returned by Imagemagick before resizing it. https://github.com/owncloud/core/issues/12465
#### The problem

You've just uploaded a 100MB TIFF file that you need to share with a restricted audience. Here is what currently happens once the file has been uploaded.
- A 36x36 thumbnail is generated in 1 minute
- Go to the Gallery, an album thumbnail might be generated in 1 minute
- Go inside the album, another thumbnail will be generated in 1 minute
- Click on the picture to get a full screen preview. 1 minute
- Go back to the Files app and share the file
- Client 1 visits the link. 1 minute to get the preview
- Client 2 with a different screen resolution visits the link. 1 minute again

You're starting to get it. Every time a slightly different preview size is asked, a full conversion of that 100MB TIFF to PNG is performed. This is highly inefficient.
#### Proposed solution
- Enforce maximum preview dimensions in the config. They're currently set to null.
- The first time we're asked to generate a preview we'll generate one of the maximum dimension indicated in the configuration and all future resizing requests will be done on that preview in order to not waste time converting the same large file over and over

Default size is 2048x2048
The maximum size preview is keeping the aspect ratio of the original and is not upscaled.
It's is immediately stored in the cache with a special `max` tag and size identifiers based on its real size.

Every time a preview of a different size is asked, we store it in the cache. This also speeds up operation at the cost of disk space.

This make a huge difference in libraries holding several large files, for people using apps such as the Gallery.

Recreated after the loss of #13582

### CODE REVIEW

**Comment by oparoz** [comment]
I lost PR #13582, so this is a copy.

cc @georgehrke

**Comment by LukasReschke** [comment]
Rebase required.

**Comment by PVince81** [comment]
This looks like a nice improvement. I can't seem to get my head around all this logic though.

**Comment by georgehrke** [comment]
Code changes itself seem fine (although I agree that it's a lot changes for one commit :smile:)
Will test asap

**Comment by oparoz** [comment]
Maybe I should have split it in 3 commits. I did move one test and 2 other methods around.

As for the logic, you have to start with
### `getPreview()`
We first check if we have a cached preview
#### `isCached()`
In order for everyone to benefit from the feature without having to re-generate all their thumbnails, we don't consider a preview to be cached until we find one with the `-max` tag

If we do, we retrieve it
#### `getCachedPreview`
* retrieves the "max" preview
* resizes the preview to fulfil the request

and if not we generate one. 
#### `generatePreview()`
* finds a provider
* generates one preview using the maximum size defined in the configuration
* stores it using the special `-max` tag. 
* resizes the preview to fulfil the request

File: lib/private/preview.php [hunk 1]
```diff
@@ -100,8 +100,8 @@ public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $max
 		$this->userView = new \OC\Files\View('/' . $user);
 
 		//set config
-		$this->configMaxX = \OC_Config::getValue('preview_max_x', null);
-		$this->configMaxY = \OC_Config::getValue('preview_max_y', null);
+		$this->configMaxX = \OC_Config::getValue('preview_max_x', 2048);
```
Comment (MorrisJobke) [inline lib/private/preview.php]: Can I ask you to use `\OCP\IConfig`?

File: lib/private/preview.php [hunk 2]
```diff
@@ -699,11 +717,128 @@ private function resizeAndCrop() {
...
+	private function generatePreview($fileId) {
+		$file = $this->getFile();
+		$preview = null;
+
+		foreach (self::$providers as $supportedMimeType => $provider) {
+			if (!preg_match($supportedMimeType, $this->mimeType)) {
+				continue;
+			}
+
+			\OC_Log::write(
```
Comment (MorrisJobke) [inline lib/private/preview.php]: Better use `\OCP\ILogger`

File: tests/lib/preview.php [hunk 3]
```diff
@@ -48,6 +48,25 @@ protected function tearDown() {
 		parent::tearDown();
 	}
 
+	public function testIsMaxSizeWorking() {
+
+		$maxX = 250;
+		$maxY = 250;
+
+		\OC_Config::setValue('preview_max_x', $maxX);
```
Comment (MorrisJobke) [inline tests/lib/preview.php]: `\OCP\IConfig`

**Comment by MorrisJobke** [comment]
@georgehrke Please have a look too

**Comment by MorrisJobke** [comment]
@georgehrke Once you're fine with this I will review.

**Comment by georgehrke** [comment]
@oparoz Isn't 2048x2048 a bit much? That's something you might need for public sharing previews on a high-resolution screen. Wouldn't 512x512 be enough for most use-cases?

**Comment by oparoz** [comment]
@georgehrke - Phones today have 2k screens, a 512x512 picture on a small, high DPI screen will look tiny. That's the main reason for me to keep it as is. And let's not forget that this is used only for full screen previews. Thumbnails will be much smaller.
It's user adjustable anyway and apps (mobile per example) could introduce settings which put another limit on previews they ask for if bandwidth is an issue.
That's my view :)

File: lib/private/preview.php (hunk 1 above)
Comment (oparoz) [inline lib/private/preview.php]: Yep, I guess, it's a good time to upgrade all those calls...

**Comment by georgehrke** [comment]
Okay, I can see your point there :)

File: lib/private/preview.php (hunk 1 above)
Comment (oparoz) [inline lib/private/preview.php]: `\OC::$server->getConfig()->getSystemValue()`

File: lib/private/preview.php (hunk 2 above)
Comment (oparoz) [inline lib/private/preview.php]: \OCP\Util::writeLog()

File: lib/private/preview.php (hunk 1 above)
Comment (georgehrke) [inline lib/private/preview.php]: or just add ifconfig to the constructor

File: lib/private/preview.php (hunk 2 above)
Comment (georgehrke) [inline lib/private/preview.php]: $this->logger->debug :D

Comment (oparoz) [inline lib/private/preview.php]: If I can touch the constructor...

File: lib/private/preview.php (hunk 1 above)
Comment (oparoz) [inline lib/private/preview.php]: That means every call to `new Preview()` will need to be updated to include new parameters or am I missing something?

Comment (MorrisJobke) [inline lib/private/preview.php]: I would go with the \OC::$server->getConfig() right inside this class in the short term.

**Comment by MorrisJobke** [comment]
Failing unit tests are unrelated

**Comment by oparoz** [comment]
Replaced by #14879

**Pull request closed without merging**